xapian-core/stemmer/KrovetzStemmer.cc

```diff
@@ -0,0 +1,6846 @@
+            ment_endings();
+            if ((dep = getdep(m_word)) != (dictEntry *)NULL) break;
+            ble_endings();
+            if ((dep = getdep(m_word)) != (dictEntry *)NULL) break;
+            ism_endings();
+            if ((dep = getdep(m_word)) != (dictEntry *)NULL) break;
+            ic_endings();
+            if ((dep = getdep(m_word)) != (dictEntry *)NULL) break;
+            ncy_endings();
+            if ((dep = getdep(m_word)) != (dictEntry *)NULL) break;
+            nce_endings();
+            dep = getdep(m_word);
+            break;
+        }
+
+        /* try for a direct mapping (allows for cases like `Italian'->`Italy' and
+            `Italians'->`Italy')
+        */
+        if (dep != (dictEntry *)NULL && dep->root[0] != '\0') {                 
+            strcpy((char *)buffer, dep->root);   
+        }
+        /* Enter into cache, at the place not used by the last cache hit */
+        if (stemCache[hval].flag == 2) {
+            strcpy(stemCache[hval].word1, term);
+            strcpy(stemCache[hval].stem1, buffer);
+            stemCache[hval].flag = 1; 
+        } else {
+            strcpy(stemCache[hval].word2, term);
+            strcpy(stemCache[hval].stem2, buffer);
+            stemCache[hval].flag = 2;
+        }
+        return (int)strlen(buffer)+1;
+    }
+
+    char * KrovetzStemmer::kstem_stemmer(char *term)
+    {
+        int length = kstem_stem_tobuffer( term, stem );
+
+        if( length )
+            return stem;
+        else
+            return term;
+    }
+
+    string KrovetzStemmer::operator()(const string & word) {
+        char thestem[80];
+        int ret;
+        ret = this->kstem_stem_tobuffer(const_cast<char*>(word.c_str()), thestem);
+        if (ret > 0) 
+            return thestem;
+        else 
+            return word;
+    }
+
+    string KrovetzStemmer::get_description() const {
+        string description="KrovetzStemmer";
+        return description;
+    }
+
+    // conflation table. static definition at end of module.
+    struct conflation_pair {
+        const char* const variant;
+        const char* const m_word;
+    };
+
+    //
+    // Below this line, you'll find hard-coded dictionaries for the 
+    // Krovetz stemmer.
+    // These arrays include the data formerly found in these files:
+    //   country_nationality.txt
+    //   dict_supplement.txt
+    //   direct_conflations.txt
+    //   exception_words.txt
+    //   head_word_list.txt
+    //   proper_nouns.txt
+    //
+
+    static const char *exceptions[] = {
+    "aide",
+    "bathe", "caste", "cute", "dame", "dime", "doge", "done", "dune",
+    "envelope", "gage", "grille", "grippe", "lobe", "mane", "mare",
+    "nape", "node", "pane", "pate", "plane", "pope", "programme",
+    "quite", "ripe", "rote", "rune", "sage", "severe", "shoppe",
+    "sine", "slime", "snipe", "steppe", "suite", "swinge", "tare",
+    "tine", "tope", "tripe", "twine",
+    NULL
```

Comment:
I realise this isn't code you wrote, but it illustrates a useful point to be aware of when working on a C/C++ library.

Arrays of `const char*` pointers like these are problematic for code that's in a shared library.

In simple terms, the strings themselves can be put into the read-only data section, but the array needs to contain addresses for those strings.  Those addresses will vary depend on the address that the shared library gets loaded at which varies[*].  So that that when a program using this shared library is run, the dynamic loader needs to fix up all the pointers in all of these arrays.  A single relocation is cheap, but en masse they can cause significant slowdowns.

The simplest fix is to make these arrays of fixed-width arrays - e.g this could instead be `static const char exceptions[][10] = ...` (where 10 is one more than the longest string).  The code using the array shouldn't need changing (unless it's doing icky things it really shouldn't be doing).

That means 10 bytes is used for every string (even though e.g. `"done"` which only needs 5, including the terminating zero byte), but this avoids a pointer per entry which is 8 bytes on a 64-bit platform (or 4 on a 32-bit platform), so using arrays would probably actually reduce the memory used a bit in this case.

See 2.4.3 in https://www.akkadia.org/drepper/dsohowto.pdf for more details (the whole document is a good read if you're interested in low-level details of how stuff works).

[*] If it didn't, every shared library would need to be allocated a span of addresses which didn't overlap with any other library - that's just not possible on a 32-bit platform as the total size of all libraries will exceed the address range; a 64-bit platform may have enough address range, but it's hard to come up with a system to uniquely allocate address ranges to libraries built by different people on different computers which may end up getting used together.

Reply:
thanks for the detailed explanation and useful information.